obs: Add doNotWarnBeforeClose setting

bobvandevijver opened the pull request:
<!--- Please fill out the following template, which will help other contributors review your Pull Request. -->

<!--- Make sure you’ve read the contribution guidelines here: https://github.com/obsproject/obs-studio/blob/master/CONTRIBUTING.rst -->

### Description
Adds a new setting to the general OBS settings, which can be checked
in order to no longer show the "confirm close" dialog when there are
still streams/recordings running.

![image](https://user-images.githubusercontent.com/1835343/99796626-1a05a680-2b2e-11eb-8147-5d944d3a1326.png)

### Motivation and Context
Sometimes, you do not want to be notified of any running streams when closing down OBS. For now, I implemented this as global setting for all streams/recordings, but this could be refined further.

See also: https://ideas.obsproject.com/posts/1201/disable-shutdown-confirmation-for-virtual-camera

### How Has This Been Tested?
Tested on my Windows 10 2009 x64 machine, by toggling the setting and virtual camera output state. This shouldn't affect any other parts as I know of.

### Types of changes
<!--- What types of changes does your PR introduce? Uncomment all that apply -->
<!--- - Bug fix (non-breaking change which fixes an issue) -->
 - New feature (non-breaking change which adds functionality) 
<!--- - Performance enhancement (non-breaking change which improves efficiency) -->
<!--- - Code cleanup (non-breaking change which makes code smaller or more readable) -->
<!--- - Breaking change (fix or feature that would cause existing functionality to change) -->
<!--- - Documentation (a change to documentation pages) -->

### Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code has been run through [clang-format](https://github.com/obsproject/obs-studio/blob/master/.clang-format).
- [x] I have read the [**contributing** document](https://github.com/obsproject/obs-studio/blob/master/CONTRIBUTING.rst).
- [x] My code is not on the master branch.
- [x] The code has been tested.
- [x] All commit messages are properly formatted and commits squashed where appropriate.
- [ ] I have included updates to all appropriate documentation.

I do not know if this needs a documentation update somewhere, but I would be more than happy to update it if I know where.

WizardCM commented:
Could you provide specific use cases for this setting?

bobvandevijver commented:
Well, the current confirmation dialog is blocking computer shutdown when anything is running (even when the virtual camera is the only active stream), which is inconvenient for our use case. Also, most applications offer an option to disable a close confirmation.

We have a couple of managed rooms with a streaming setup, where the user interaction with the setup must be as simple as possible. We use OBS (started minimized on login) and control it using a [StreamDeck](https://www.elgato.com/en/gaming/stream-deck). The user should never need the possibility to even see the main OBS window.

We also want to be able to use the video input in other applications (such as Microsoft Teams), which is why we want to use the virtual camera. It can be started automatically with OBS using the correct flag (I might want to add those flags to the settings as well), but that also means that we always have an active output running, blocking our shutdown.

So, for "automated" environment, where a shutdown must always just shutdown OBS, the blocking dialog is quite inconvenient.

matoi974 commented:
I have the same issue, I always start obs-studio with replay buffer, and it's annoying to confirm close, I even often kill it in task manager instead. Definitely won't hurt to have this option.

bobvandevijver commented:
Is there anything I can do to get this merged so it can be included in the next release?

bobvandevijver commented:
@jp9000 / @WizardCM Could anyone take a look at including this in OBS 27? Is there anything I can do?

WizardCM commented:
The OBS 27 feature freeze was back in late March. We're in the process of finalising the remaining bug fixes for features introduced in 27, then once 27 is out to the masses we'll have a couple of weeks fixing any bugs the wider userbase finds, and then after that we'll begin looking at Enhancement PRs again.

bobvandevijver commented:
@WizardCM Any chance this PR can be merged for a next release?

Warchamp7 commented:
I'd like to see this moved to the Advanced section under General with a wording change

"Shut down all active inputs on exit"

WizardCM commented:
Additionally the commit prefix should be UI: not obs:

bobvandevijver commented:
Alright, I've rebased the changes on the current master, adjusted the settings name, its location and fixed the commit prefix!

![image](https://user-images.githubusercontent.com/1835343/130071163-2c39998d-33fb-47c3-a33c-82c2f7dd8b4d.png)

bobvandevijver commented on UI/window-basic-settings.cpp:
```diff
@@ -553,6 +553,7 @@ OBSBasicSettings::OBSBasicSettings(QWidget *parent)
 	HookWidget(ui->reconnectRetryDelay,  SCROLL_CHANGED, ADV_CHANGED);
 	HookWidget(ui->reconnectMaxRetries,  SCROLL_CHANGED, ADV_CHANGED);
 	HookWidget(ui->processPriority,      COMBO_CHANGED,  ADV_CHANGED);
+	HookWidget(ui->shutDownActiveOutputsOnExit, CHECK_CHANGED,  ADV_CHANGED);
```
I was wondering about this one, is this okay or do you have a suggestion for this?

jp9000 commented:
Here's my take on it:

`Shut down all active outputs on exit` is a confusing name. The name of the widget should be `confirmOnExit`, that way it's super short and simple. Then the text for it should just be something like `Show confirmation dialog when exiting the program while active`. The option should be enabled by default. Additionally, the option should be in the "Output" group rather than the "General" group.

bobvandevijver commented:
I agree the widget name is too long so I adjusted it to `confirmOnExit`.

OBS currently shows "OBS is currently active. All streams/recordings will be shut down. Are you sure you wish to exit?". I guess the translation should try to follow that? That is what I tried with the latest version anyways

Warchamp7 commented:
> @Warchamp7 I would really like to have this included in the next release so we can start using this in our university college rooms in September. Anything else that I can do in order to achieve that?

Heya, we discussed this a bit more on Discord. I think this looks good with the checkbox default on, but a suggested wording change to "Show active outputs warning on exit"

The current string you have is a bit lengthy and will probably get even longer in some translations.

Fenrirthviti commented:
Unfortunately, we don't accept translations through direct PRs, they need to be submitted to our [Crowdin page](http://crowdin.com/project/obs-studio)

Please remove the Dutch translation (you can resubmit it back to the Crowdin once this is merged, however) and squash the commits, and this should be ready for final review.

bobvandevijver commented:
Done!

dodgepong commented:
Can you squash this into a single commit?

bobvandevijver commented:
Sure, also done. (although I actually assumed you would let Github do that during the actual merge)

Fenrirthviti commented:
It makes the final review easier to have everything that's supposed to be there in a single commit for review. Thanks!
